include/ldsCtrlEst_h/lds_mpcctrl.h

```diff
@@ -0,0 +1,533 @@
+                            const Matrix& xr, size_t n_sim);
+
+  /**
+   * @brief     Calculate the trajectory for the simulation step
+   *
+   * @param     x0  The initial state
+   * @param     u0  The initial control input
+   * @param     xr  The reference trajectory
+   * @param     out Print out step information
+   *
+   * @return    The trajectory for the simulation step
+   */
+  OSQPSolution* slow_update(const Vector& x0, const Vector& u0,
+                            const Matrix& xr, size_t n_sim);
+
+  /**
+   * @brief     Create an identity matrix with an offset axis
+   * 
+   * @param     n The size of the matrix
+   * @param     k The offset axis
+   * 
+   * @return    The identity matrix with an offset axis
+   */
+  arma::SpMat<data_t> eye_offset(size_t n, int k = -1) {
+    arma::SpMat<data_t> mat(n, n);
+    int start = (k < 0) ? -k : 0;
+    for (int i = start; i < n && i + k < n; i++) {
+      mat(i, i + k) = 1;
+    }
+    return mat;
+  }
+
+  /**
+   * @brief     Create a block diagonal matrix given two input matrices
+   * 
+   * @param     m1 The first matrix
+   * @param     m2 The second matrix
+   * 
+   * @return    The block diagonal matrix
+   */
+  Matrix block_diag(Matrix m1, Matrix m2) {
+    // Calculate the total rows and columns of the block diagonal matrix
+    size_t rows = m1.n_rows + m2.n_rows;
+    size_t cols = m1.n_cols + m2.n_cols;
+
+    // Create the block diagonal matrix
+    Matrix bd = arma::zeros<Matrix>(rows, cols);
+    bd.submat(0, 0, m1.n_rows - 1, m1.n_cols - 1) = m1;
+    bd.submat(m1.n_rows, m1.n_cols, rows - 1, cols - 1) = m2;
+
+    return bd;
+  }
+
+  // void print_osqp_csc_matrix(const OSQPCscMatrix* matrix) {
+  //   std::cout << "[" << matrix->m << ", " << matrix->n << "]:\n";
+
+  //   for (int j = 0, k = 0; j < (matrix->p[matrix->n]); j++) {
+  //     while (k <= matrix->n && matrix->p[k + 1] <= j) k++;
+  //     std::cout << "(" << matrix->i[j]  << ", " << k << ") " << matrix->x[j] << "\n";
+  //   }
+  //   std::cout << std::endl;
+  // }
+
+  void Init() {
+    A_ = sys_.A();
+    B_ = sys_.B();
+    n_ = B_.n_rows;
+    m_ = B_.n_cols;
+    xi_ = arma::zeros<Vector>(n_);
+    t_sim_ = 0;
+  }
+};
+
+// Implement methods
+
+template <typename System>
+inline MpcController<System>::MpcController(const System& sys, Vector u_lb,
+                                            Vector u_ub)
+    : sys_(sys), lb_(u_lb), ub_(u_ub) {
+  Init();
+}
+
+template <typename System>
+inline MpcController<System>::MpcController(System&& sys, Vector u_lb,
+                                            Vector u_ub)
+    : sys_(std::move(sys)), lb_(u_lb), ub_(u_ub) {
+  Init();
+}
+
+template <typename System>
+Vector MpcController<System>::Control(data_t t_sim, const Vector& x0,
+                                      const Vector& u0, const Matrix& xr) {
+  size_t n_sim = t_sim / sys_.dt();  // Number of points per simulation step
+
+  OSQPSolution* sol;
```

Comment:
Does the optimization problem get reinitialized on every step? That will probably be a lot slower than if parts can be reused and just the state is updated

Reply:
Yes, I think the problem does get reinitialized every step. I mostly was trying to stick to what the Python package had (except currently the fast update function is never called, which I haven't fixed yet because the slow update doesn't work correctly), but I can look more into how to keep some of the previous states to prevent reinitializing at every step.